### ISSUE TITLE
LspRename and LspDocumentFormat produce odd results in Python and C++

### ISSUE THREAD

**Issue by d3rhans** [prabirshrestha/vim-lsp#216]
Good afternoon,

I'm trying vim-lsp since yesterday and so far the two features I'm most interested in (Autocompletion and diagnostics) seem to work pretty well.

Nevertheless, I've encountered issues with LspRename and LspDocumentFormat, both working with C++ files as well as Python files.

**Python:**

Starting from this simple example:

    import numpy as np
  
    class Foo:
        def __init__(self, test):
            self.foo = test
  
        def foo(self):
            a = np.zeros((9))
            print(a)
            print(np.asarray(self.foo))
  
    test = Foo([1, 2, 2, 3])

If I have the cursor somewhere on foo in self.foo, execute LspRename and change its name to `bar`, the change is executed as expected, but also the last line is changed to:

    test = Foo([1, 2, 2, 3])est = Foo([1, 2, 2, 3])

The same happens when I execute LspDocumentFormat. 

**C++**

Starting from:

    #include <iostream>

    struct Foo {
        int a;  
        int b;  
 
        void bar() { 
            std::cout << a << " " << b;
        }       
    };      
 
    int main()
    {       
        int bar = 12;

        std::cout << "Hallo Welt!" << std::endl;
 
        std::cout << bar << std::endl;
 
        Foo f{.a=12, .b=13}; 
        f.bar();
 
        f.bar();
 
        return 0;
    }

If I try to rename the method Foo::bar to foo, I get:

    #include <iostream>
  
    struct Foo {
        int a;
        int b;
  
        void foo) {
            std::cout << a << " " << b;
        }   
    };
  
    int main()
    {
        int bar = 12; 

        std::cout << "Hallo Welt!" << std::endl;
  
        std::cout << bar << std::endl;
  
        Foo f{.a=12, .b=13};
        f.foo);
  
        f.foo);
  
        return 0;
    }

So the rename is done where it should, but a '(' gets lost in the process. If I perform LspDocumentFormat on the same C++ code, the result is completely messed up:

    #include <iostream>

    struct Foo {
      nt a;
      nt b;

      oid bar() { td::cout << a << " " << b; 
    };

    int main() 
      nt bar = 12;

      td::cout << "Hallo Welt!" << std::endl;

      td::cout << bar << std::endl;

      oo f{.a = 12, .b = 13};
      .bar();

      .bar();

      eturn 0;
    }

I use: 
- vim 8.1 (Patches: 1-549) as shipped with Fedora Linux 29
- pyls 0.21.4 with all optional extensions installed
- clangd 7.0.0

My vimrc is currently mostly clear (apart from vim-lsp, asyncomplete and ayncomplete-lsp Plugin setup and their configuration as suggested in their documentation).

I've attached the four log-files for the actions provided. If you need any further details, just let me know.

Thank you in advance.

[pyls_rename.log](https://github.com/prabirshrestha/vim-lsp/files/2691259/pyls_rename.log)
[pyls_format.log](https://github.com/prabirshrestha/vim-lsp/files/2691260/pyls_format.log)
[clangd_rename.log](https://github.com/prabirshrestha/vim-lsp/files/2691261/clangd_rename.log)
[clangd_format.log](https://github.com/prabirshrestha/vim-lsp/files/2691262/clangd_format.log)

**Comment by kadrach** [comment]
Same issue, here is an example: https://asciinema.org/a/GQkahx9I36CnXXvP9ZuYBApgX

Notice how, the second time I call `LspDocumentFormat`, the cursor and content are on the last line of the file.

In this example my vim configuration is empty besides vim-lsp and it's dependencies as well as language servers. This happens with all configured language servers for me (python, ruby, vue, ...).

**Comment by yskripachov** [comment]
Same here (debian9|ubuntu18.04). Still have to check end_character to manage LspDocumentFormat.
[https://github.com/yskripachov/vim-lsp/commit/268028558fb2741c3632677651a3d798e9268646](url)

**Comment by d3rhans** [comment]
@yskripachov I've added the check in your commit and it seems to fix the problems with the python example. C++ however still gets messed up as before.

**Comment by yskripachov** [comment]
@d3rhans Commit was tested with pyls that responses on /rename with one range with the whole file content. Mostly likely clangd returns several ranges and looks like apply_text_edits function is replacing extra char in selection. Could you try to set selection=exclusive as it was before Format #209?

**Comment by yskripachov** [comment]
For me - it works. Both LspDocumentFormat and LspRename. But pyls returns (at least for small files) the whole file content. clangd judging by log returns several ranges and eats last char in selection. So not sure that LspRename will work with C++ example.

**Comment by mattn** [comment]
Thank your confirm. So my pull-request fix this issue?

**Comment by yskripachov** [comment]
I can not check it with clangd language server.

**Comment by mattn** [comment]
@d3rhans Could you please check #218 ?

**Comment by yskripachov** [comment]
pyls and clangd works for me with [https://github.com/yskripachov/vim-lsp/commit/ef23401ef07ea6e25c7d4399352e4016a8498219](https://github.com/yskripachov/vim-lsp/commit/ef23401ef07ea6e25c7d4399352e4016a8498219)

**Comment by d3rhans** [comment]
@mattn I've checked #218, python looks good, C++ has the same results as before. With the additional changes suggested by @yskripachov both, the C++ and Python example work (Rename as well as document format).

Only thing I noticed, in the C++ file the intendation width is changed from 4 to 2. Is this related to the vim settings (I currently have the default: tabstop=8, shiftwidth=8, noexpandtab), clangd or vim-lsp?

Thank you.
